x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessAuthenticationService.java

```diff
@@ -148,7 +148,19 @@ public AuthenticationService getAuthenticationService() {
     }
 
     private void validate(final RemoteAccessAuthentication remoteAccessAuthentication) {
-        final Subject effectiveSubject = remoteAccessAuthentication.getAuthentication().getEffectiveSubject();
+        final Authentication authentication = remoteAccessAuthentication.getAuthentication();
+        authentication.checkConsistency();
```

Comment:
The only way in which I can think to test this is to have a spy-ed instance here that throws an exception on `checkConsistency`. That we could at least assert that `RemoteAccessAuthenticationService` behaves correctly in the case of an error here.

Reply:
I considered this approach but it does _not_ solve the core problem that we are _not_ able to assert that `checkConsistency` is actually called. To call `checkConsistency`, we first need an `Authentication` object. But we cannot because assertion is enabled in tests and the assertion calls `checkConsistency` during instantiation and fails. So the whole thing still fails early on before it can even enter the validator. By extracting a validator, we can get some value by asserting the validator is called during authentication. But since the validator itself cannot be asserted, I think it's not really worth the effort.

Another alternative that I have considered is to *not* rely on `Assertions.ENABLED` but instead introducing a similar control specific for the Authentication class and make it configurable for tests, basically equivalent to dropping the `final` modifier for `Assertions.ENABLED`. The test can then disable the assertion to successfully instantiate an Authentication object, then call `checkConsistency` and expect an failure. It still feels an overkill for me. But please let me know if you think otherwise.

Reply:
> To call checkConsistency, we first need an Authentication object. But we cannot because assertion is enabled in tests and the assertion calls checkConsistency during instantiation and fails.

I was thinking something along the lines of:

```java
final Authentication mocked = spy(AuthenticationTestHelper.builder().build());
doThrow(new IllegalArgumentException("consistency failure")).when(mocked).checkConsistency();
expectThrows(..., () -> validator.validate(new RemoteAccessAuthentication(..., mocked)));
```

But that would require to drop the `final` modifier on the `Authentication` class and is quite a lot setup to test a very specific code path that is already largely exercised by the existing code (we test that other validation errors fail authentication already).

I don't think it's worth doing, just wanted to further explain my original suggestion.

Reply:
Thanks for the explanation. Yes dropping the `final` modifier would make the test work. I also agree it does not seem to justify the change given the effort went into making the authentication final.

I think part of the problem is also because we only run assertion in constructors instead of throwing runtime error. If we could just throw runtime error in constructor, there will be no need to test this validation process separately. But that means adding a bit overhead to the constructor and potentially (hopefully not) behaviour change. I think we can discuss this option. But for now, I'll just push this through as is. Thanks!